### PR TITLE
BrushY: A dummy vertical Brush for scaling/panning the YAxis

### PR DESCRIPTION
**Pull request by kerguler**
The changes clone and transform Brush as a vertical slider, but do not cast a function to it.

## Description

The new BrushY keeps track of the startIndex and endIndex within a user-defined number of steps. BrushY can be linked to the YAxis to set an appropriate domain for vertical scrolling/zooming.

## Related Issue

Zoom and pan - Issue #710 : https://github.com/recharts/recharts/issues/710#top

## Motivation and Context

Brush performs horizontal scrolling/zooming by filtering the source data as needed. BrushY removes the filtering step and provides a dummy vertical slider to scale/pan the YAxis.

## How Has This Been Tested?

A demo is included in the demo/component folder to show how the link between BrushY and YAxis can be established by the user. The demo uses the useState hook to update the YAxis limits. In addition, a test is included in the test/cartesian folder (similar to the test for Brush) to confirm that the component is properly rendered.

To enable rendering the component and remove the data (domain) filtering step, changes are also done in generateCategoricalChart.tsx.

## Screenshot:

Screenshot 2024-01-03 at 08 04 19

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ x ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ x ] My code follows the code style of this project.
- [ x ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ x ] I have added tests to cover my changes.
- [ ] I have added a storybook story or extended an existing story to show my changes
- [ x ] All new and existing tests passed.

### CODE REVIEW

File: .eslintrc
Comment (ckifer) [inline .eslintrc:80]: Looks like these changes are unrelated to this PR, can you remove changes to the lint file?

File: demo/component/BrushYDemo.tsx
Comment (ckifer) [inline demo/component/BrushYDemo.tsx:21]: We don't really use the demo project anymore, can you add this code to the storybook instead?

File: package.json
Comment (ckifer) [inline package.json:3]: Please don't bump the version unless an official recharts release, we'll handle this internally

**Review by ckifer — CHANGES_REQUESTED** [review]
Really appreciate you taking the time to add this, I think we'll need to think on this a little more before full release though. We want to sure we test fully to catch any edge cases and aren't adding to our already large pile of tech debt. 

There are some changes in this PR that are outside of the context of the PR (version bump, eslint changes) that will need to be reverted if we are to merge this. 

Will also approvals on direction from some others as this doesn't look like the direction we want to continue to move in. 

Really appreciate the legwork!

File: package.json
Comment (kerguler) [reply package.json:3]: Of course, I'll revert this.

File: .eslintrc
Comment (kerguler) [reply .eslintrc:80]: I included these as I was receiving errors while compiling locally. I will revert them.

File: demo/component/BrushYDemo.tsx
Comment (kerguler) [reply demo/component/BrushYDemo.tsx:21]: Of course, as soon as I figure out how to do it :)

**Comment by kerguler** [comment]
Thank you @ckifer for your lightning fast response. I need this component for the react app we are developing. If you decide not to include it in the master, could you please kindly advise the best approach to achieve this functionality? At the moment, I am keeping the changes in the fork and installing them using yalc.

File: .eslintrc
Comment (ckifer) [reply .eslintrc:80]: Hmm odd, with your code or with other code? The current code in the codebase shouldn't have any errors with the lint settings besides maybe in the demo project (which we don't use 😅)

File: demo/component/BrushYDemo.tsx
Comment (ckifer) [reply demo/component/BrushYDemo.tsx:21]: Should be similar code luckily! You can add this to the Brush examples: storybook - Examples - Cartesian - Brush

**Comment by ckifer** [comment]
We definitely want the functionality eventually! We're just trying to take a different approach with some things (as you can see the generate function is very large and hard to follow). If we can get this to a good place we can maybe merge it, but we might also wait until after some more refactoring has been done to make it easier to understand/test and so we can have a more robust API. So it just might be a little bit before this gets merged/released with that said. 

I'll wait for your changes and some reviews from others and we can go from there. 

In terms of publishing yourself you can publish your own fork of recharts on npm if you'd like though it's always better to follow along with master since you'll miss out. You could publish your fork and use that until we merge this if you really need this functionality now. 

Thanks!

File: src/chart/generateCategoricalChart.tsx
Comment (PavelVanecek) [inline src/chart/generateCategoricalChart.tsx:1974]: Yes this is precisely the pattern we are trying to get rid of 😬 try to think of an API that does not require cloning, and replacing the props internally

Comment (PavelVanecek) [reply src/chart/generateCategoricalChart.tsx:1974]: Hm on the other hand it would make sense to have the same API as the existing Brush. After all it's the same thing, isn't it.

Comment (kerguler) [reply src/chart/generateCategoricalChart.tsx:1974]: @PavelVanecek I couldn't find an easy answer to the problem of rendering the same component twice. Now that I am a bit more familiar with the workings of the package, it might have been easier than I thought. However, I had to make some detailed conformational changes, and played with the endIndex range. So, a separate component made sense

File: .eslintrc
Comment (kerguler) [reply .eslintrc:80]: Neither. The problem occurs when I try to commit with git. It won't let me unless the lint issues are resolved.

File: demo/component/BrushYDemo.tsx
Comment (kerguler) [reply demo/component/BrushYDemo.tsx:21]: I'll add it as a new component as I couldn't figure out how to incorporate changes in the Brush component and have it rendered twice. I trust that the planned refactoring will enable this. Until then...

File: .eslintrc
Comment (ckifer) [reply .eslintrc:80]: I think it's because of the changes to the demo components though, lint staged it linting when you commit and the lint errors are from the demo files

Comment (kerguler) [reply .eslintrc:80]: [STARTED] Preparing lint-staged...
[SUCCESS] Preparing lint-staged...
[STARTED] Running tasks for staged files...
[STARTED] package.json — 7 files
[STARTED] *.{ts,tsx,js,jsx} — 3 files
[STARTED] eslint --fix
[FAILED] eslint --fix [FAILED]
[FAILED] eslint --fix [FAILED]
[FAILED] eslint --fix [FAILED]
[STARTED] Applying modifications from tasks...
[SKIPPED] Skipped because of errors from tasks.
[STARTED] Reverting to original state because of errors...
[SUCCESS] Reverting to original state because of errors...
[STARTED] Cleaning up temporary files...
[SUCCESS] Cleaning up temporary files...

✖ eslint --fix:

/Users/kamil/Projects/VEClim/recharts/src/cartesian/BrushY.tsx
  308:39  error  Unexpected empty arrow function                          @typescript-eslint/no-empty-function
  310:39  error  Unexpected empty arrow function                          @typescript-eslint/no-empty-function
  543:27  error  'isSlideMoving' is assigned a value but never used       @typescript-eslint/no-unused-vars
  543:42  error  'isTravellerMoving' is assigned a value but never used   @typescript-eslint/no-unused-vars
  543:61  error  'isTravellerFocused' is assigned a value but never used  @typescript-eslint/no-unused-vars

✖ 5 problems (5 errors, 0 warnings)

husky - pre-commit hook exited with code 1 (error)

Comment (kerguler) [reply .eslintrc:80]: I'll fix these and see if it works

Comment (ckifer) [reply .eslintrc:80]: Oh yeah those are actual lint errors that need fixed

Comment (kerguler) [reply .eslintrc]: Okay, it committed without any errors and with BrushYDemo intact (might just be my luck). Just to note, the part I included in the storybook does nothing. It is the demo that connects BrushY with YAxis. I am not sure how to have the same functionality in the storybook, but it might not be necessary. Hance the reason to name it as a "dummy" component.

File: src/cartesian/BrushY.tsx
Comment (ckifer) [inline src/cartesian/BrushY.tsx:1]: I didn't even see this file... I wonder if we can do these changes without a new component and instead with an `orientation` prop or similar

Comment (kerguler) [reply src/cartesian/BrushY.tsx:1]: The problem is not being able to render the same component twice. If that's solved, than BrushY could be integrated into Brush

File: src/chart/generateCategoricalChart.tsx
Comment (nikolasrieble) [reply src/chart/generateCategoricalChart.tsx:1974]: I believe this does not add any extra effort for future migration, if it simply stays in line with the existing brush. We will have to migrate the brush either way.

File: src/cartesian/BrushY.tsx
Comment (nikolasrieble) [inline src/cartesian/BrushY.tsx:1]: This file is certainly the blocking thing here. 
We do not want to simply copy code with small variations. 

This will reduce our velocity going forward massively. 
Please do reuse the existing code for the Brush and extend it.

**Review by nikolasrieble — CHANGES_REQUESTED** [review]
Great work with user value here, please do not copy paste the component but instead extend the existing one in alignment with the libraries concepts. 

i.e. we also have a CartesianAxis and then we have an XAxis component and a YAxis component. 

That is a pattern that seems well applicable here. What do you think?

File: src/cartesian/BrushY.tsx
Comment (nikolasrieble) [reply src/cartesian/BrushY.tsx:1]: The pattern to solve this is done for the Axis. A shared CartesianAxis exists, but then we also have XAxis and YAxis. Both of them internally use CartesianAxis.

File: src/cartesian/BrushY.tsx
Comment (kerguler) [reply src/cartesian/BrushY.tsx:1]: I agree that the existing pattern makes sense. Would it be acceptable to decommission Brush and create both an XBrush and a YBrush component?

Comment (ckifer) [reply src/cartesian/BrushY.tsx:1]: To be completely honest, in the next (or a future) major version I think we should remove Brush entirely and instead implement a recharts native pan and zoom feature. Brush is clunky, especially for Y values. 

(That's opinion only and probably won't happen)

What's the issue with removal of the once prop and just using Brush with an orientation prop?

Comment (nikolasrieble) [reply src/cartesian/BrushY.tsx:1]: Also possible, but I like the consistency of patterns between components that we would have if we had YBrush and XBrush

Comment (ckifer) [reply src/cartesian/BrushY.tsx:1]: Sure, lets get it to that state where components are shared and go from there

**Comment by cesarzxk** [comment]
I believe that the brush could be accepted as a child of the x and y axes, behaving in the orientation of the axis.

**Comment by ckifer** [comment]
Yeah I think that could make sense. Breaking API change of course. Once we get brush info into context all of this gets easier

**Comment by crazy4chrissi** [comment]
I find this super useful for vertical barcharts (`layout="vertical"`). There, I would expect the BrushY to act similar as the Brush for horizontal barcharts, i.e. control the range of dimension values. However, the default `onChange` does not do this because BrushY does not call `this.handleBrushChange`. Would be cool if this would be called by default for vertical barcharts.

**Comment by ckifer** [comment]
I'm going to close this as stale, but lets revisit in 3.1 or later

**Pull request closed without merging**
